[NFC][LoopVectorize] Cache result of requiresScalarEpilogue

david-arm opened the pull request:
Caching the decision returned by requiresScalarEpilogue means that
we can avoid printing out the same debug many times, and also
avoids repeating the same calculation. This function will get more
complex when we start to reason about more early exit loops, such
as in PR #88385.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1388,32 +1388,42 @@ class LoopVectorizationCostModel {
+    if (CachedResult)
+      return *CachedResult;
+
+    auto NeedsScalarEpilogue = [&](bool IsVectorizing) -> bool {
```
I don't think the closure is necessary here. You could initialize Res to false here, then if `isScalarEpilogueAllowed()` evaluate the other conditions and set Res to true instead of returning true.

fhahn commented:
Would it be simpler to have a separate helper `setRequiresScalarEpilogue` that's called once up-front with `requiresScalarEpilogue` simply returning the decision as done in multiple other places?

david-arm commented:
> Would it be simpler to have a separate helper `setRequiresScalarEpilogue` that's called once up-front with `requiresScalarEpilogue` simply returning the decision as done in multiple other places?

I'm happy to do it this way, but I couldn't convince myself of the earliest place to put such code that is guaranteed not to crash. I can go off and have a look of course, but if you happen to have any ideas that would be great!

david-arm commented:
> Would it be simpler to have a separate helper `setRequiresScalarEpilogue` that's called once up-front with `requiresScalarEpilogue` simply returning the decision as done in multiple other places?

OK, I've tried doing this, but I realised there are places where we have to invalidate the decision due to changes in the scalar epilogue status or interleave groups. However, that also means my first version was also incorrect even though all tests passed!

fhahn reviewed (commented):
> > Would it be simpler to have a separate helper `setRequiresScalarEpilogue` that's called once up-front with `requiresScalarEpilogue` simply returning the decision as done in multiple other places?
> 
> OK, I've tried doing this, but I realised there are places where we have to invalidate the decision due to changes in the scalar epilogue status or interleave groups. However, that also means my first version was also incorrect even though all tests passed!


Thanks for checking, does this mean we are missing a new test to cover this case?

david-arm commented:
> > > Would it be simpler to have a separate helper `setRequiresScalarEpilogue` that's called once up-front with `requiresScalarEpilogue` simply returning the decision as done in multiple other places?
> > 
> > 
> > OK, I've tried doing this, but I realised there are places where we have to invalidate the decision due to changes in the scalar epilogue status or interleave groups. However, that also means my first version was also incorrect even though all tests passed!
> 
> Thanks for checking, does this mean we are missing a new test to cover this case?

So for the interleave info case I couldn't find any possible scenario where requireScalarEpilogue would change value after invalidating the interleave groups. For example if you have an interleave group with gaps in a block that needs predication and you happen to support masked interleaved accesses, then we'll require a scalar epilogue. But the place that invalidates the interleave info later on requires that you *don't* support masked interleaved accesses.

However, I did find an issue with my original patch when changing the scalar epilogue lowering status. I realised we have a missing test case where we request predication on a loop that requires a scalar epilogue (due to an early exit). We should correctly refuse to tail-fold and fall back on normal vectorisation and jump to the scalar epilogue for the last iteration. This test was broken with my previous patch, which I've now fixed.

david-arm commented:
Rebase + fix code formatting

david-arm committed:
Add test for requesting tail-folding on a loop requiring a scalar epilogue

There is a flag attached to the loop that requests tail-folding,
but this cannot be honoured because the early exit requires a
scalar epilogue. So we should fall back on normal vectorisation
with a scalar epilogue.

david-arm committed:
[NFC][LoopVectorize] Cache result of requiresScalarEpilogue

Caching the decision returned by requiresScalarEpilogue means that
we can avoid printing out the same debug many times, and also
avoids repeating the same calculation. This function will get more
complex when we start to reason about more early exit loops, such
as in PR #88385. The only problem with this is we sometimes have to
invalidate the previous result due to changes in the scalar epilogue
status or interleave groups.
